### ISSUE TITLE
Using Accessors in filters

### ISSUE THREAD

**Issue by liakos21** [abbasudo/laravel-purity#4]
Hello, thanks for this nice package.

I have a Product model that has two columns for pricing.

- sale_price
- regular_price

Then, I am using an accessor named 'price' that returns the regular_price if no sale_price is set.

When I am trying to use the following filter function: `?filters[price][$gte]=1` I am getting the following error:
`Column not found: 1054 Unknown column 'price' in 'where clause` I know that Query Builder does not have access to accessors and that is why this is happening but I think there is a way to get over it. I am just not experienced/confident enough to provide a good solution. My guess is that a raw DB query would solve this.

**Comment by abbasudo** [comment]
Yes, you are correct, the Query Builder doesn't have access to accessors in Laravel.
there are many solutions to your problem. ill start with the easiest one:
you can get around this problem by simply using `$or` in your query params. so it will be something like this :
`?filters[$or][1][regular_price][$gte]=1&filters[$or][2][sale_price][$gte]=1`


the second way to solve the issue would be to create a custom filter and write the following logic:
```php
class priceGteFilter extends Filter
{
    
    protected static string $operator = '$priceGte';
    
    public function apply(): Closure
    {
        return function (Builder $query) {
            foreach ($this->values as $value) {
                $query->whereRaw("IFNULL(sale_price, regular_price) >= {$value}");
            }
        };
    }
}
```
after that, you can call the API by this params: `?filters[$priceGte]=1`. this solution is **cleaner for frontend** users however you have to make a custom filter for each method you want to filter.

the third way of solving this is by adding an actual price column to the table and ensuring that the column is filled by using SQL triggers. to do this you have to change your migration file to this :
```php
    public function up(): void
    {
        Schema::create('books', function (Blueprint $table) {
            $table->id();
            $table->string('title');
            $table->unsignedInteger('regular_price');
            $table->unsignedInteger('price');
            $table->unsignedInteger('sale_price')->nullable();
            $table->timestamps();
        });

        DB::unprepared('
            CREATE TRIGGER insert_price
            BEFORE INSERT ON products
            FOR EACH ROW
            BEGIN
                SET NEW.price = IFNULL(NEW.sale_price, NEW.regular_price);
            END;
            ');

        DB::unprepared('
            CREATE TRIGGER update_price
            BEFORE UPDATE ON products
            FOR EACH ROW
            BEGIN
                SET NEW.price = IFNULL(NEW.sale_price, NEW.regular_price);
            END;
            ');
    }
```
personally, I would prefer the last solution because someday ill need to query that price. however, SQL triggers are no longer popular and nobody likes to use them ([link](https://blog.sqlauthority.com/2007/05/24/sql-server-disadvantages-problems-of-triggers/), [link](https://www.red-gate.com/simple-talk/databases/sql-server/database-administration-sql-server/sql-server-triggers-good-scary/)).
